### PR TITLE
Implement dynamic param passing for RDS per env, tune prod params

### PR DESCRIPTION
**Pull request by njdister**
This allows us to more easily tune RDS parameters per environment.  This also tunes 4 parameters in prod CCS RDS that were much lower than their HealthApt equivalents.

Side note: Do we need import mode any longer?  Would be great to rip it out if unneeded.  It's unclear to me whether it even helped with the bulk imports at all.

### CODE REVIEW

**Comment by njdister** [comment]
Terraform plan prod:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.stateful.aws_db_parameter_group.default_mode will be updated in-place
  ~ resource "aws_db_parameter_group" "default_mode" {
        arn         = "arn:aws:rds:us-east-1:577373831711:pg:bfd-prod-default-mode-parameter-group"
        description = "Sets parameters for standard operation"
        family      = "postgres9.6"
        id          = "bfd-prod-default-mode-parameter-group"
        name        = "bfd-prod-default-mode-parameter-group"
        tags        = {}

      + parameter {
          + apply_method = "immediate"
          + name         = "random_page_cost"
          + value        = "1"
        }
      + parameter {
          + apply_method = "immediate"
          + name         = "temp_buffers"
          + value        = "8192"
        }
      + parameter {
          + apply_method = "immediate"
          + name         = "work_mem"
          + value        = "32768"
        }
      + parameter {
          + apply_method = "pending-reboot"
          + name         = "max_worker_processes"
          + value        = "16"
        }
    }

  # module.stateful.module.master.aws_db_instance.db will be updated in-place
  ~ resource "aws_db_instance" "db" {
        address                               = "bfd-prod-master.clyryngdhnko.us-east-1.rds.amazonaws.com"
        allocated_storage                     = 16000
        apply_immediately                     = false
        arn                                   = "arn:aws:rds:us-east-1:577373831711:db:bfd-prod-master"
        auto_minor_version_upgrade            = true
        availability_zone                     = "us-east-1b"
        backup_retention_period               = 21
        backup_window                         = "05:00-06:00"
        ca_cert_identifier                    = "rds-ca-2015"
        copy_tags_to_snapshot                 = true
        db_subnet_group_name                  = "bfd-prod-subnet-group"
        deletion_protection                   = true
        enabled_cloudwatch_logs_exports       = [
            "postgresql",
            "upgrade",
        ]
        endpoint                              = "bfd-prod-master.clyryngdhnko.us-east-1.rds.amazonaws.com:5432"
        engine                                = "postgres"
        engine_version                        = "9.6.11"
        final_snapshot_identifier             = "bfd-prod-master-final"
        hosted_zone_id                        = "Z2R2ITUGPM61AM"
        iam_database_authentication_enabled   = false
        id                                    = "bfd-prod-master"
        identifier                            = "bfd-prod-master"
        instance_class                        = "db.r5.24xlarge"
      ~ iops                                  = 16000 -> 80000
        kms_key_id                            = "arn:aws:kms:us-east-1:577373831711:key/82901607-504c-40b0-85bc-3be929ca10e6"
        license_model                         = "postgresql-license"
        maintenance_window                    = "fri:07:00-fri:08:00"
        max_allocated_storage                 = 0
        monitoring_interval                   = 15
        monitoring_role_arn                   = "arn:aws:iam::577373831711:role/rds-monitoring-role"
        multi_az                              = false
        name                                  = "bfdtemp"
        option_group_name                     = "default:postgres-9-6"
        parameter_group_name                  = "bfd-prod-default-mode-parameter-group"
        password                              = (sensitive value)
        performance_insights_enabled          = false
        performance_insights_retention_period = 0
        port                                  = 5432
        publicly_accessible                   = false
        replicas                              = [
            "bfd-prod-replica2",
            "bfd-prod-replica3",
            "bfd-prod-replica1",
        ]
        resource_id                           = "db-7V7HUAVDF5JP5W5S3APVNN4WDA"
        security_group_names                  = []
        skip_final_snapshot                   = false
        status                                = "available"
        storage_encrypted                     = true
        storage_type                          = "io1"
        tags                                  = {
            "Environment" = "prod"
            "Layer"       = "data"
            "application" = "bfd"
            "business"    = "oeda"
            "role"        = "master"
            "stack"       = "prod"
        }
        username                              = "bfduser"
        vpc_security_group_ids                = [
            "sg-0a2fb33487c23a233",
            "sg-0a6734c1551266727",
            "sg-0aee7caa9ab3c0914",
            "sg-0c86a4b1fdd035130",
        ]
    }

  # module.stateful.module.replica1.aws_db_instance.db will be updated in-place
  ~ resource "aws_db_instance" "db" {
        address                               = "bfd-prod-replica1.clyryngdhnko.us-east-1.rds.amazonaws.com"
        allocated_storage                     = 16000
        apply_immediately                     = false
        arn                                   = "arn:aws:rds:us-east-1:577373831711:db:bfd-prod-replica1"
        auto_minor_version_upgrade            = true
        availability_zone                     = "us-east-1a"
        backup_retention_period               = 0
        backup_window                         = "05:00-06:00"
        ca_cert_identifier                    = "rds-ca-2015"
        copy_tags_to_snapshot                 = true
        db_subnet_group_name                  = "bfd-prod-subnet-group"
        deletion_protection                   = true
        enabled_cloudwatch_logs_exports       = [
            "postgresql",
            "upgrade",
        ]
        endpoint                              = "bfd-prod-replica1.clyryngdhnko.us-east-1.rds.amazonaws.com:5432"
        engine                                = "postgres"
        engine_version                        = "9.6.11"
        final_snapshot_identifier             = "bfd-prod-replica1-final"
        hosted_zone_id                        = "Z2R2ITUGPM61AM"
        iam_database_authentication_enabled   = false
        id                                    = "bfd-prod-replica1"
        identifier                            = "bfd-prod-replica1"
        instance_class                        = "db.r5.24xlarge"
      ~ iops                                  = 16000 -> 80000
        kms_key_id                            = "arn:aws:kms:us-east-1:577373831711:key/82901607-504c-40b0-85bc-3be929ca10e6"
        license_model                         = "postgresql-license"
        maintenance_window                    = "fri:07:00-fri:08:00"
        max_allocated_storage                 = 0
        monitoring_interval                   = 15
        monitoring_role_arn                   = "arn:aws:iam::577373831711:role/rds-monitoring-role"
        multi_az                              = false
        name                                  = "bfdtemp"
        option_group_name                     = "default:postgres-9-6"
        parameter_group_name                  = "bfd-prod-default-mode-parameter-group"
        performance_insights_enabled          = false
        performance_insights_retention_period = 0
        port                                  = 5432
        publicly_accessible                   = false
        replicas                              = []
        replicate_source_db                   = "bfd-prod-master"
        resource_id                           = "db-M736V6Y7WBRSRAGQ6WU22YRCG4"
        security_group_names                  = []
        skip_final_snapshot                   = false
        status                                = "available"
        storage_encrypted                     = true
        storage_type                          = "io1"
        tags                                  = {
            "Environment" = "prod"
            "Layer"       = "data"
            "application" = "bfd"
            "business"    = "oeda"
            "role"        = "replica1"
            "stack"       = "prod"
        }
        username                              = "bfduser"
        vpc_security_group_ids                = [
            "sg-054d0623e58df94de",
            "sg-0a2fb33487c23a233",
            "sg-0a6734c1551266727",
            "sg-0aee7caa9ab3c0914",
        ]
    }

  # module.stateful.module.replica2.aws_db_instance.db will be updated in-place
  ~ resource "aws_db_instance" "db" {
        address                               = "bfd-prod-replica2.clyryngdhnko.us-east-1.rds.amazonaws.com"
        allocated_storage                     = 16000
        apply_immediately                     = false
        arn                                   = "arn:aws:rds:us-east-1:577373831711:db:bfd-prod-replica2"
        auto_minor_version_upgrade            = true
        availability_zone                     = "us-east-1b"
        backup_retention_period               = 0
        backup_window                         = "05:00-06:00"
        ca_cert_identifier                    = "rds-ca-2015"
        copy_tags_to_snapshot                 = true
        db_subnet_group_name                  = "bfd-prod-subnet-group"
        deletion_protection                   = true
        enabled_cloudwatch_logs_exports       = [
            "postgresql",
            "upgrade",
        ]
        endpoint                              = "bfd-prod-replica2.clyryngdhnko.us-east-1.rds.amazonaws.com:5432"
        engine                                = "postgres"
        engine_version                        = "9.6.11"
        final_snapshot_identifier             = "bfd-prod-replica2-final"
        hosted_zone_id                        = "Z2R2ITUGPM61AM"
        iam_database_authentication_enabled   = false
        id                                    = "bfd-prod-replica2"
        identifier                            = "bfd-prod-replica2"
        instance_class                        = "db.r5.24xlarge"
      ~ iops                                  = 16000 -> 80000
        kms_key_id                            = "arn:aws:kms:us-east-1:577373831711:key/82901607-504c-40b0-85bc-3be929ca10e6"
        license_model                         = "postgresql-license"
        maintenance_window                    = "fri:07:00-fri:08:00"
        max_allocated_storage                 = 0
        monitoring_interval                   = 15
        monitoring_role_arn                   = "arn:aws:iam::577373831711:role/rds-monitoring-role"
        multi_az                              = false
        name                                  = "bfdtemp"
        option_group_name                     = "default:postgres-9-6"
        parameter_group_name                  = "bfd-prod-default-mode-parameter-group"
        performance_insights_enabled          = false
        performance_insights_retention_period = 0
        port                                  = 5432
        publicly_accessible                   = false
        replicas                              = []
        replicate_source_db                   = "bfd-prod-master"
        resource_id                           = "db-36LTZ4USZTY2ATGZCVYHJ6P7YU"
        security_group_names                  = []
        skip_final_snapshot                   = false
        status                                = "available"
        storage_encrypted                     = true
        storage_type                          = "io1"
        tags                                  = {
            "Environment" = "prod"
            "Layer"       = "data"
            "application" = "bfd"
            "business"    = "oeda"
            "role"        = "replica2"
            "stack"       = "prod"
        }
        username                              = "bfduser"
        vpc_security_group_ids                = [
            "sg-054d0623e58df94de",
            "sg-0a2fb33487c23a233",
            "sg-0a6734c1551266727",
            "sg-0aee7caa9ab3c0914",
        ]
    }

  # module.stateful.module.replica3.aws_db_instance.db will be updated in-place
  ~ resource "aws_db_instance" "db" {
        address                               = "bfd-prod-replica3.clyryngdhnko.us-east-1.rds.amazonaws.com"
        allocated_storage                     = 16000
        apply_immediately                     = false
        arn                                   = "arn:aws:rds:us-east-1:577373831711:db:bfd-prod-replica3"
        auto_minor_version_upgrade            = true
        availability_zone                     = "us-east-1c"
        backup_retention_period               = 0
        backup_window                         = "05:00-06:00"
        ca_cert_identifier                    = "rds-ca-2015"
        copy_tags_to_snapshot                 = true
        db_subnet_group_name                  = "bfd-prod-subnet-group"
        deletion_protection                   = true
        enabled_cloudwatch_logs_exports       = [
            "postgresql",
            "upgrade",
        ]
        endpoint                              = "bfd-prod-replica3.clyryngdhnko.us-east-1.rds.amazonaws.com:5432"
        engine                                = "postgres"
        engine_version                        = "9.6.11"
        final_snapshot_identifier             = "bfd-prod-replica3-final"
        hosted_zone_id                        = "Z2R2ITUGPM61AM"
        iam_database_authentication_enabled   = false
        id                                    = "bfd-prod-replica3"
        identifier                            = "bfd-prod-replica3"
        instance_class                        = "db.r5.24xlarge"
      ~ iops                                  = 16000 -> 80000
        kms_key_id                            = "arn:aws:kms:us-east-1:577373831711:key/82901607-504c-40b0-85bc-3be929ca10e6"
        license_model                         = "postgresql-license"
        maintenance_window                    = "fri:07:00-fri:08:00"
        max_allocated_storage                 = 0
        monitoring_interval                   = 15
        monitoring_role_arn                   = "arn:aws:iam::577373831711:role/rds-monitoring-role"
        multi_az                              = false
        name                                  = "bfdtemp"
        option_group_name                     = "default:postgres-9-6"
        parameter_group_name                  = "bfd-prod-default-mode-parameter-group"
        performance_insights_enabled          = false
        performance_insights_retention_period = 0
        port                                  = 5432
        publicly_accessible                   = false
        replicas                              = []
        replicate_source_db                   = "bfd-prod-master"
        resource_id                           = "db-XBXMSXOJFJFWIUKWTADYPKG6JE"
        security_group_names                  = []
        skip_final_snapshot                   = false
        status                                = "available"
        storage_encrypted                     = true
        storage_type                          = "io1"
        tags                                  = {
            "Environment" = "prod"
            "Layer"       = "data"
            "application" = "bfd"
            "business"    = "oeda"
            "role"        = "replica3"
            "stack"       = "prod"
        }
        username                              = "bfduser"
        vpc_security_group_ids                = [
            "sg-054d0623e58df94de",
            "sg-0a2fb33487c23a233",
            "sg-0a6734c1551266727",
            "sg-0aee7caa9ab3c0914",
        ]
    }

Plan: 0 to add, 5 to change, 0 to destroy.
```

**Comment by njdister** [comment]
Based on the plan output, it looks like the IOPS bump in prod did not go through?

**Comment by RickHawesUSDS** [comment]
> Based on the plan output, it looks like the IOPS bump in prod did not go through?

I had my doubts that we can bump IOPS independently of storage size, but I'm not an expert.

File: ops/terraform/env/prod/stateful/main.tf [hunk 1]
```diff
@@ -16,6 +16,13 @@ module "stateful" {
...
   }
 
+  db_params = [
+    {name="max_worker_processes", value="16", apply_on_reboot=true},
+    {name="random_page_cost", value="1", apply_on_reboot=false},
+    {name="temp_buffers", value="8192", apply_on_reboot=false},
+    {name="work_mem", value="32768", apply_on_reboot=false}
+  ]
+
   db_import_mode = {
```
Comment (RickHawesUSDS) [inline ops/terraform/env/prod/stateful/main.tf]: Cool way to do this.

**Comment by RickHawesUSDS** [comment]
https://severalnines.com/database-blog/architecture-and-tuning-memory-postgresql-databases

**Comment by RickHawesUSDS** [comment]
Took some time to read about Postgres memory tuning. Here's one that seemed like a good starting point. https://severalnines.com/database-blog/architecture-and-tuning-memory-postgresql-databases

It suggests that we should increase the shared_buffers to 8GB and the effective_cache_size to 50% of memory 192GB. As well as monitor the logs for temp files and increase the work_mem size if we find them.

File: ops/terraform/env/prod/stateful/main.tf (hunk 1 above)
Comment (karlmdavis) [reply ops/terraform/env/prod/stateful/main.tf]: Also: I don't think there's any reason to only apply these changes in prod; as long as our DB HW is the same in all environments (it is now), we likely want the same config in all environments.

**Comment by karlmdavis** [comment]
> Took some time to read about Postgres memory tuning. Here's one that seemed like a good starting point. https://severalnines.com/database-blog/architecture-and-tuning-memory-postgresql-databases
> 
> It suggests that we should increase the shared_buffers to 8GB and the effective_cache_size to 50% of memory 192GB. As well as monitor the logs for temp files and increase the work_mem size if we find them.

@RickHawesUSDS We actually *can't* tune those values on RDS: Amazon sets them for us, based on the instance type. If we disagree with their choices, we basically have to run PostgreSQL ourselves.

What @njdister is doing here is tuning the other settings that we actually can tweak.

See this part of a conversation that Nathan and I had, for reference: https://cmsgov.slack.com/archives/CJQRH9RLG/p1570478909257300?thread_ts=1570477474.254000&cid=CJQRH9RLG

**Comment by RickHawesUSDS** [comment]
> > Took some time to read about Postgres memory tuning. Here's one that seemed like a good starting point. https://severalnines.com/database-blog/architecture-and-tuning-memory-postgresql-databases
> > It suggests that we should increase the shared_buffers to 8GB and the effective_cache_size to 50% of memory 192GB. As well as monitor the logs for temp files and increase the work_mem size if we find them.
> 
> @RickHawesUSDS We actually _can't_ tune those values on RDS: Amazon sets them for us, based on the instance type. If we disagree with their choices, we basically have to run PostgreSQL ourselves.
> 
> What @njdister is doing here is tuning the other settings that we actually can tweak.
> 
> See this part of a conversation that Nathan and I had, for reference: https://cmsgov.slack.com/archives/CJQRH9RLG/p1570478909257300?thread_ts=1570477474.254000&cid=CJQRH9RLG

Good to know.

File: ops/terraform/env/prod/stateful/main.tf (hunk 1 above)
Comment (njdister) [reply ops/terraform/env/prod/stateful/main.tf]: > Also: I don't think there's any reason to only apply these changes in prod; as long as our DB HW is the same in all environments (it is now), we likely want the same config in all environments.

RDS instances in prod-sbx (db.m5.2xlarge) are a good bit smaller than in test or prod (db.r5.24xlarge)

File: ops/terraform/env/test/stateful/main.tf [hunk 2]
```diff
@@ -16,6 +16,8 @@ module "stateful" {
     allocated_storage = 12000
   }
 
+  db_params = []
```
Comment (karlmdavis) [inline ops/terraform/env/test/stateful/main.tf]: Let's have this mirror `prod`, since it's the same HW.

File: ops/terraform/env/prod/stateful/main.tf [hunk 3]
```diff
@@ -16,6 +16,13 @@ module "stateful" {
     allocated_storage = 16000
   }
 
+  db_params = [
+    {name="max_worker_processes", value="16", apply_on_reboot=true},
```
Comment (karlmdavis) [inline ops/terraform/env/prod/stateful/main.tf]: Can we go with the PGTune recommendations for this one and friends?

```
max_worker_processes = 96
max_parallel_workers_per_gather = 48
max_parallel_workers = 96
```

File: ops/terraform/env/prod/stateful/main.tf [hunk 4]
```diff
@@ -16,6 +16,13 @@ module "stateful" {
...
+    {name="random_page_cost", value="1", apply_on_reboot=false},
```
Comment (karlmdavis) [inline ops/terraform/env/prod/stateful/main.tf]: I think we should also set this related PGTune recommendation for SAN disks:

```
effective_io_concurrency = 300
```

But if you want to do that as part of a separate PR, so we can benchmark it, that's cool.

File: ops/terraform/env/prod/stateful/main.tf [hunk 5]
```diff
@@ -16,6 +16,13 @@ module "stateful" {
...
+    {name="temp_buffers", value="8192", apply_on_reboot=false},
```
Comment (karlmdavis) [inline ops/terraform/env/prod/stateful/main.tf]: Isn't this already the default? https://www.postgresql.org/docs/9.6/runtime-config-resource.html

File: ops/terraform/env/prod/stateful/main.tf [hunk 6]
```diff
@@ -16,6 +16,13 @@ module "stateful" {
...
+    {name="work_mem", value="32768", apply_on_reboot=false}
```
Comment (karlmdavis) [inline ops/terraform/env/prod/stateful/main.tf]: FYI: We currently use 40 DB connections per app server (overkill, but whatev) * max-ASG-size of 24 * 32MB = 30.7 GB RAM used for `work_mem`.

Which is fine! Just wanted to note it.

File: ops/terraform/env/prod/stateful/main.tf [hunk 7]
```diff
@@ -16,6 +16,13 @@ module "stateful" {
...
+  ]
```
Comment (karlmdavis) [inline ops/terraform/env/prod/stateful/main.tf]: Can we also bump up how much data is collected by `ANALYZE`?

```
default_statistics_target = 1000
```

File: ops/terraform/env/prod/stateful/main.tf (hunk 5 above)
Comment (njdister) [reply ops/terraform/env/prod/stateful/main.tf]: The unit is in chunks of 8k, so the default is 1024 x 8k = 8M.  This bumps it up to 64M.  The postgres doc happily leaves that out, and its very all over the place which values are in which chunks.  Fun times.

File: ops/terraform/env/prod/stateful/main.tf (hunk 6 above)
Comment (njdister) [reply ops/terraform/env/prod/stateful/main.tf]: Wont this only be per AZ, so 8/24 instances in that AZ will connect to their respective replica.

File: ops/terraform/env/prod/stateful/main.tf (hunk 4 above)
Comment (njdister) [reply ops/terraform/env/prod/stateful/main.tf]: I keep waffling on this one due to a huge lack of publicly posted data but I found a bit more that indicates this should be ok on IO1 type storage, though can be less performant on the less reliable GP2 storage.

File: ops/terraform/env/prod/stateful/main.tf (hunk 6 above)
Comment (karlmdavis) [reply ops/terraform/env/prod/stateful/main.tf]: Good call!
